### PR TITLE
Start adding support for more Leaf cameras

### PR DESCRIPTION
**Pull request by pedrocr**
I've started to support more MOS (and IIQ) format cameras from Leaf. There are three new cases:
- a little-endian version of the basic MOS format that's working fine (aptusII-5)
- several LJpeg cameras that don't work with the current LJpeg code (all the other aptus cameras)
- a few cameras that use a specific encoding and need special tiff parsing (the credo's and the ixr). These are not working but I did lift the specific code from dcraw and left it there for future reference

I've gotten a file for each of the cameras for testing at: 

http://pedrocr.pt/leaf/ 

Here is a per-camera breakdown:
- aptus22: works fine, the original camera I had an example file for
- aptusII-10: "Out of buffer read" in the LJpeg code
- aptusII-10R: "LJpegDecompressor: Quantized components not supported."
- aptusII-12: "LJpegDecompressor: Quantized components not supported."
- aptusII-5: added support in this pull request
- aptusII-7: "Out of buffer read" in the LJpeg code
- aptusII-8: "Corrupt JPEG data: bad Huffman code:17"
- credo40: requires the PhaseOneC decoding and specific tiff parsing code
- credo60: requires the PhaseOneC decoding and specific tiff parsing code
- credo80: requires the PhaseOneC decoding and specific tiff parsing code
- ixr: requires the PhaseOneC decoding and specific tiff parsing code

Klaus, do you want to have a look at the LJpeg stuff? As for the special tiff parsing do you want to have a look at dcraw (parse_phase_one) and tell me if normal tiff parsing can be used or if I should just write a one-off parser? I tried parsing it with TiffIFD but it didn't work.

### CODE REVIEW

**Comment by pedrocr** [comment]
This should be ready to merge now. I've now added support for the PhaseOneC cameras (Credo 40, Credo 60/IXR and Credo 80). The Ljpeg ones are still broken but that's not in this code anyway.

File: RawSpeed/MosDecoder.cpp [hunk 1]
```diff
@@ -68,34 +73,147 @@ void MosDecoder::parseXMP(TiffEntry *xmp) {
...
+    char *text = (char *) meta->getDataWrt();
+    uint32 size = meta->count;
+    text[size-1] = 0; //Make sure the data is NUL terminated so that scanf never reads beyond limits
+
+    // dcraw does actual parsing, since we just want one field we bruteforce it
+    char *neutobj = (char *) memmem(text, size, "NeutObj_neutrals", 16);
+    if (neutobj) {
+      uint32 tmp[4] = {0};
+      sscanf((const char *)neutobj+44, "%u %u %u %u", &tmp[0], &tmp[1], &tmp[2], &tmp[3]);
+      mRaw->metadata.wbCoeffs[0] = (float) tmp[0]/tmp[1];
```
Comment (klauspost) [inline RawSpeed/MosDecoder.cpp]: You need to check if tmp[1] to tmp[3] are zero.

File: RawSpeed/MosDecoder.cpp [hunk 2]
```diff
@@ -68,34 +73,147 @@ void MosDecoder::parseXMP(TiffEntry *xmp) {
...
   this->checkCameraSupported(meta, make, model, "");
 }
 
 void MosDecoder::decodeMetaDataInternal(CameraMetaData *meta) {
   setMetaData(meta, make, model, "", 0);
+
+  // Fetch the white balance (see dcraw.c parse_mos for more metadata that can be gotten)
+  if (mRootIFD->hasEntryRecursive(LEAFMETADATA)) {
+    TiffEntry *meta = mRootIFD->getEntryRecursive(LEAFMETADATA);
+    char *text = (char *) meta->getDataWrt();
```
Comment (klauspost) [inline RawSpeed/MosDecoder.cpp]: It shouldn't be necessary to get a write pointer.

File: RawSpeed/MosDecoder.cpp [hunk 3]
```diff
@@ -68,34 +73,147 @@ void MosDecoder::parseXMP(TiffEntry *xmp) {
...
+    char *neutobj = (char *) memmem(text, size, "NeutObj_neutrals", 16);
+    if (neutobj) {
+      uint32 tmp[4] = {0};
+      sscanf((const char *)neutobj+44, "%u %u %u %u", &tmp[0], &tmp[1], &tmp[2], &tmp[3]);
+      mRaw->metadata.wbCoeffs[0] = (float) tmp[0]/tmp[1];
+      mRaw->metadata.wbCoeffs[1] = (float) tmp[0]/tmp[2];
+      mRaw->metadata.wbCoeffs[2] = (float) tmp[0]/tmp[3];
+    }
+    if (text)
+      delete text;
```
Comment (klauspost) [inline RawSpeed/MosDecoder.cpp]: 'text' should be freed by TiffEntry, thus this is probably a double free.

File: RawSpeed/MosDecoder.cpp [hunk 4]
```diff
@@ -28,24 +28,29 @@ namespace RawSpeed {
...
-  delete xmpText;
 }
 
 void MosDecoder::parseXMP(TiffEntry *xmp) {
   if (xmp->count <= 0)
     ThrowRDE("MOS Decoder: Empty XMP");
 
-  xmpText = xmp->getDataWrt();
+  uchar8 *xmpText = xmp->getDataWrt();
   xmpText[xmp->count - 1] = 0; // Make sure the string is NUL terminated
```
Comment (klauspost) [inline RawSpeed/MosDecoder.cpp]: If you use xmp->getString(), this isn't needed.

File: RawSpeed/MosDecoder.cpp (hunk 2 above)
Comment (klauspost) [inline RawSpeed/MosDecoder.cpp]: If you use xmp->getString, this will get simpler.

**Comment by klauspost** [comment]
Had a look through it. Nothing major.

**Comment by pedrocr** [comment]
I've reworked the string handling. Code should be much cleaner now.

**Comment by klauspost** [comment]
Looks great!

**Pull request merged**